src/feature.js

```diff
@@ -1,147 +1,186 @@
-    console.log("initialized");
+  async handleActivated(activeInfo) {
+    const tab = await browser.tabs.get(activeInfo.tabId);
+    this.activatedTabId = tab.id;
+    this.maybeUpdateSettingListener(tab.id, tab.url);
   }
 
-  /** handleButtonClick
-   *
-   * - instrument browserAction button clicks
-   * - change label
-   */
-  handleButtonClick() {
-    console.log("handleButtonClick");
-    // note: doesn't persist across a session, unless you use localStorage or similar.
-    this.timesClickedInSession += 1;
-    console.log("got a click", this.timesClickedInSession);
-    browser.browserAction.setBadgeText({
-      text: this.timesClickedInSession.toString(),
-    });
+  async handleUpdated(tabId, changeInfo, tabInfo) {
+    if (!changeInfo.url) {
+      return;
+    }
 
-    // telemetry: FIRST CLICK
-    if (this.timesClickedInSession === 1) {
-      browser.study.sendTelemetry({ event: "button-first-click-in-session" });
+    const url = changeInfo.url;
+    console.log(`tab update : TabId: ${tabId}, URL changed to ${url}`);
+    this.maybeUpdateSettingListener(tabId, url);
+    if (!isSupportedURLProtocol(url)) {
+      return;
     }
 
-    // telemetry EVERY CLICK
-    browser.study.sendTelemetry({
-      event: "button-click",
-      timesClickedInSession: "" + this.timesClickedInSession,
-    });
+    const pageId = await getHostNameSaltedHash(url);
+    this.feature.update("visitPage", pageId);
+  }
 
-    // webExtension-initiated ending for "used-often"
-    //
-    // - 3 timesClickedInSession in a session ends the study.
-    // - see `../Config.jsm` for what happens during this ending.
-    if (this.timesClickedInSession >= 3) {
-      browser.study.endStudy("used-often");
+  handleRemoved(tabId, removeInfo) {
+    if (this.activatedTabId === tabId &&
+        browser.autoplay.autoplaySettingChanged.hasListener(this.settingListener)) {
+      browser.autoplay.autoplaySettingChanged.removeListener(this.settingListener);
     }
   }
 }
 
+class Feature {
+  constructor() {}
+
+  configure(studyInfo) {
+    const feature = this;
+    const { variation, isFirstRun } = studyInfo;
+    console.log(studyInfo);
+
+    this.tabsMonitor = new TabsMonitor();
+    this.tabsMonitor.configure(this);
+    const sendPingIntervalMS = 1 * 24 * 60 * 60 * 1000;
+
+    this.sendPingsScheduler = setInterval(() => {
+      browser.autoplay.sendTelemetry();
+    }, sendPingIntervalMS);
+
+    browser.autoplay.setPreferences(variation.name);
+  }
+
+  update(type, data) {
+    if (type === "visitPage" || type === "autoplayOccur") {
+      data = {pageId: data};
+    }
+    browser.autoplay.updatePingData(type, data);
+  }
+
+  /**
+   * Called at end of study, and if the user disables the study or it gets
+   * uninstalled by other means.
+   */
+  async cleanup() {
+    this.tabsMonitor.clear();
+    // restoring to the default option.
+    browser.autoplay.setPreferences("allow-and-remember");
```

Comment:
Hmm, this is assuming that `browser.autoplay.setPreferences("allow-and-remember");` will always reset to the default prefs. That assumption could become false if your team decides to uplift a patch that changes the default pref.

To be certain, I would recommend going the extra mile and adding a `browser.autoplay.clearPreferences` function.